storage/transfermanager/downloader.go

```diff
@@ -0,0 +1,149 @@
+package transfermanager
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+)
+
+type TransferManagerOption interface {
+	apply(*transferManagerConfig)
+}
+
+func WithWorkers(numWorkers int) TransferManagerOption {
+	return &withWorkers{numWorkers: numWorkers}
+}
+
+type withWorkers struct {
+	numWorkers int
+}
+
+func (ww withWorkers) apply(tm *transferManagerConfig) {
+	tm.numWorkers = ww.numWorkers
+}
+
+func WithPartSize(partSize int) TransferManagerOption {
+	return &withPartSize{partSize: partSize}
+}
+
+type withPartSize struct {
+	partSize int
+}
+
+func (wps withPartSize) apply(tm *transferManagerConfig) {
+	tm.partSize = wps.partSize
+}
+
+// etc for additional options.
+
+type transferManagerConfig struct {
+	// Workers in thread pool; default numCPU/2 based on previous benchmarks?
+	numWorkers int
+	// Size of shards to transfer; Python found 32 MiB to be good default for
+	// JSON downloads but gRPC may benefit from larger.
+	partSize int
+	// Timeout for a single operation (including all retries).
+	perOperationTimeout time.Duration
+
+	// others, including opts that apply to uploads.
+}
+
+// etc
+
+type Downloader struct {
+	ctx             context.Context
+	client          *storage.Client
+	config          *transferManagerConfig
+	objectInputs    []DownloadObjectInput    // Will be sent via a channel to workers
+	directoryInputs []DownloadDirectoryInput // This does obj listing/os.File calls and then converts to object inputs
+	output          <-chan *DownloadOutput   // Channel for completed downloads; used to feed results iterator
+	done            <-chan bool              // Used to signal completion of all downloads.
+	// etc
+}
+
+// Create a new Downloader to add operations to.
+// Choice of transport, etc is configured on the client that's passed in.
+func NewDownloader(c *storage.Client, opts ...TransferManagerOption) (*Downloader, error) {
+
+	return nil, nil
+}
+
+// Input for a single object to download.
+type DownloadObjectInput struct {
+	// Required fields
+	Bucket      string
+	Source      string
+	Destination io.WriterAt
+
+	// Optional fields
+	Generation     *int64
+	Conditions     *storage.Conditions
+	EncryptionKey  []byte
+	Offset, Length int64 // if specified, reads only a range.
```

Comment:
So this will be different than the regular behaviour (I believe normally a length of 0 will work at least for HTTP but will not read the whole object), or do I make these (at least length) pointers as well?